src/llama-graph.cpp

```diff
@@ -204,7 +204,10 @@ void llm_graph_input_cls::set_input(const llama_ubatch * ubatch) {
         std::vector<int> target_pos(n_seqs_unq, -1);
         std::vector<int> target_row(n_seqs_unq, -1);
 
-        bool last = cparams.pooling_type == LLAMA_POOLING_TYPE_LAST;
+        const bool last = (
+             cparams.pooling_type == LLAMA_POOLING_TYPE_LAST ||
+            (cparams.pooling_type == LLAMA_POOLING_TYPE_RANK && arch == LLM_ARCH_QWEN3) // qwen3 reranking & embedding models use last token
```

Comment:
I am wondering if it makes sense to remove pooling type RANK all together from `libllama`? Do you have any thoughts about if having a separate pooling class `RANK` is really necessary?

Reply:
I think you could get really close to merging RANK with LAST. The main differentiator is in `llm_graph_context::build_pooling` where you apply `cls_out` to map from the last token of the last hidden state to the classification output (usual yes/no). Unlike the other pooling types, you actually need knowledge of the model weights to do the calculation.